cmd-ref: plots: flexible plots docs (#3691)
iterative/dvc.org: content/docs/command-reference/plots/index.md

# plots

A set of commands to visualize and compare _plot data_:
[show](/doc/command-reference/plots/show),
[diff](/doc/command-reference/plots/diff),
[modify](/doc/command-reference/plots/modify), and
[templates](/doc/command-reference/plots/templates).

## Synopsis

```usage
usage: dvc plots [-h] [-q | -v] {show,diff,modify,templates} ...

positional arguments:

>   {show,diff,modify,templates}

Reviewer: Should be this here I think?
```suggestion
  COMMAND
```

>   {show,diff,modify,templates}

Reviewer: Should be this here I think?
```suggestion
  COMMAND
```

>                         Use `dvc plots CMD --help` to display command-specific help.

Reviewer: No need for this line in the refs.
```suggestion
```

    show                Generate plots from target files or plots definitions from `dvc.yaml` file.
    diff                Show multiple versions of plot data by plotting them in a single image.
    modify              Modify display properties of data-series plot outputs (has no effect on image-type plots).

>     templates           Write built-in plots templates to a directory (.dvc/plots by default).

Reviewer 2: would be better to have `dvc plots templates` (show available templates, including custom?), `dvc plots template --dump <template-name>` dumps it. 

(feels like dumping a template is an aux command compared to show / diff  and even modify)

Author: We can discuss and modify the usage, but I would prefer to avoid it in this PR.

> 
> optional arguments:
>   -h, --help            show this help message and exit
>   -q, --quiet           Be quiet.
>   -v, --verbose         Be verbose.

Reviewer: Or any of these...
```suggestion
```

    diff                Show multiple versions of plot data by plotting them in a single image.
    modify              Modify display properties of data-series plot outputs (has no effect on image-type plots).
    templates           Write built-in plots templates to a directory (.dvc/plots by default).

optional arguments:
  -h, --help            show this help message and exit
  -q, --quiet           Be quiet.
  -v, --verbose         Be verbose.
```

> ## Types of metrics

Reviewer 2: do we want to move it somewhere? (not sure if we have a good place)

Reviewer: This is in https://dvc.org/doc/command-reference/metrics#types-of-metrics now. At least link from `plots` ref(s), yeah.

Author: As I mentioned in the [issue description](https://github.com/iterative/dvc.org/pull/3691#issue-1283821901) I don't think we should be trying to explain plots in terms of metrics types. `dvc plots` is a visualization that can be created from any type of viable data. Yes, we have a convenient output type that lets DVC pick up plots without defining them but I believe saying that its special type of metrics makes users think about it in a particular way that is unnecessary.

Reviewer 2: Agreed, and as I mentioned I'm not sure we have a good place for this.

Reviewer 2: @jorgeorpinel let's drop it in the second place as well?

# plots

> A set of commands to visualize and compare _plot data_:

Reviewer: We're not using term "plot data" anywhere else. Should we be specific instead?

```suggestion
A set of commands to visualize and compare data series or images from ML projects:
```
> Needs wrapping

positional arguments:
  COMMAND
    show        Generate plots from target files or plots definitions from `dvc.yaml` file.
    diff        Show multiple versions of plot data by plotting them in a single image.

>     modify      Modify display properties of data-series stage plots (has no effect on image-type plots).

Reviewer: Can we wrap these texts please? It's a good practice to always sanity check changes [on the review app](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/command-reference/plots):

![image](https://user-images.githubusercontent.com/1477535/178393917-ad6ced17-8ca7-4538-8169-4bda0e8e2168.png)

Reviewer: OK how about this
```suggestion
    show      Generate plots from target files or from `plots`
                definitions in `dvc.yaml`.
    diff      Show multiple versions of a plot by overlaying them
                in a single image.
    modify    Modify display properties of data-series plots
                defined in stages (has no effect on image plots).
```

positional arguments:
  COMMAND
    show         Generate plot from a metrics file.
    diff         Plot differences in metrics between commits.
    modify       Modify display properties of data-series plots (has no effect on image-type plots).
    show         Generate plots from target files or plots definitions from `dvc.yaml` file.
    diff         Show multiple versions of plot data by plotting them in a single image.
    modify       Modify display properties of data-series plot outputs (has no effect on image-type plots).
```

## Types of metrics

DVC has two concepts for metrics, that represent different results of machine

> learning training or data processing:

Reviewer: Should we also kill or rephrase this section in the `metrics` ref. since we're now explaining `plots` as a data-related command? As well as other `plots` links from that ref.

See https://dvc.org/doc/command-reference/metrics#types-of-metrics and 3rd paragraph in the **Description** below that.

Author: I agree. Should we just remove it? It seems to me that it doesn't make sense to keep this paragraph.

Reviewer: > Should we just remove it?

Yes please @pared

> DVC provides a set of commands to visualize data produced by machine learning
> experiments. Usual plot examples are AUC curves, loss functions, confusion
> matrices, among others.

Reviewer: I agree to present `plots` mainly as a data-centric command instead of another type of "metric" (which was always a little confusing). But one use case is still to use `plots` to visualize "2D" pipeline metrics right? So I think we can elaborate a little in this description:

```suggestion
DVC provides a set of commands to visualize data produced by machine learning
projects. Usual plots include AUC curves, loss functions, or confusion
matrices, for example.

Plots are a great alternative to `dvc metrics` when working with multi-dimensional
performance data. They also help you present and compare [experiments] effectively.
```

☝🏼 `[experiments]` link missing

> This type of data is created by users, or generated by user data processing
> code.
> 

Reviewer: No longer needed (redundant) I think.
```suggestion
```

## Types of plots

> `dvc plots` is able to visualize two types of data:

Reviewer: I'd roll this change back, or at least avoid adding more structure. The **Description** should not need more H2/3s. If it does it's a sign that we need a new User Guide on top of this change...

1. Data series files, which can be JSON, YAML, CSV or TSV.
2. Image files in JPEG, GIF, or PNG format.

> DVC generates visualizations as static HTML webpages that can be open with a web
> browser. They can also be saved as SVG or PNG image files from the browser.

Reviewer: Rolling back, it was clearer before 🙂 
```suggestion
DVC generates plots as static HTML webpages that can be open with a web browser.
They can also be saved as SVG or PNG image files from the browser.
```

Data-series plots utilize [Vega-Lite](https://vega.github.io/vega-lite/) for
rendering (declarative JSON grammar for defining graphics). Images are rendered
using `<img>` tags directly.

> ## Supported file formats

Reviewer: Hmmm... I see we already have all [this structure](https://dvc.org/doc/command-reference/plots#supported-file-formats) in. I'll make an issue about flattening this/ extracting to some guide ⏳  For now most of these should probably be H3s (contained under the **Description**)?

Reviewer: > I'll make an issue about flattening this/ extracting to some guide 

A closely-related epic already exists so I just updated that one: https://github.com/iterative/dvc.org/issues/2572

Author: > Hmmm... I see we already have all [this structure](https://dvc.org/doc/command-reference/plots#supported-file-formats) in. I'll make an issue about flattening this/ extracting to some guide ⏳ For now most of these should probably be H3s (contained under the Description)?

You mean `Definig a plot too`?

positional arguments:
  COMMAND
    show      Generate plots from target files or from `plots`
                definitions in `dvc.yaml`.
    diff      Show multiple versions of a plot by overlaying them
                in a single image.
    modify    Modify display properties of data-series plots
                defined in stages (has no effect on image plots).
    templates   Write built-in plots templates to a directory (.dvc/plots by default).
```

## Description

DVC provides a set of commands to visualize data produced by machine learning
projects. Usual plots include AUC curves, loss functions, or confusion matrices,
for example. Plots are a great alternative to `dvc metrics` when working with
multi-dimensional performance data. They also help you present and compare

> [[experiments]](https://github.com/iterative/dvc.org/pull/3691#pullrequestreview-1024053543) effectively.

Reviewer: Oops
```suggestion
[experiments] effectively.
```

positional arguments:
  COMMAND
    show        Generate plots from target files or from `plots`
                definitions in `dvc.yaml`.
    diff        Show multiple versions of a plot by overlaying them
                in a single image.
    modify      Modify display properties of data-series plots
                defined in stages (has no effect on image plots).
    templates   Write built-in plots templates to a directory

>                 (.dvc/plots by default).

Reviewer: `plots` sub-command descriptions

Author: This and following suggestions has been used to update https://github.com/iterative/dvc/pull/7972

}
```

> ## Definining a plot

Reviewer 3: "Configuring a plot"? WDYT?

It might help here to explain that you can specify a data file and configure the plot in the plots commands. Adding a plot to dvc.yaml saves you from having to repeat it and allows you to consistently view a specified set of plots. dvc.yaml can also persist your plots configuration and enables you to combine data from multiple files into one plot.

@jorgeorpinel Any thoughts/ideas on wording?

Author: > enables you to combine data from multiple files into one plot.

This is pretty much in the next paragraph... should I add it here too?

Reviewer 3: Sorry @pared, my focus on this comment was on the benefits of putting plots configuration into `dvc.yaml`, regardless of whether they are standalone or stage plots.

On second thought, I think we should include `dvc.yaml` in the section title ("Configuring in `dvc.yaml`"?). Although plots may be configured from the CLI, neither the data sources nor the configuration get saved, and only a subset of plots configuration is actually supported in the current CLI (for example, plots with multiple data sources can't be configured in the CLl).

WDYT?

Reviewer: > "Configuring a plot"? WDYT?

Sorry for the delay @dberenbaum (I didn't want to add yet another reviewer... but here we are).

Actually I think "defining" is better. But I'm about to make a suggestion on what to do with all these very long blocks of new content (may need another PR split TBH).

In order to create visualizations, users need to provide the data and
(optionally) configuration that will help customize the plot. DVC provides two
ways to configure visualizations. Users can mark specific <abbr>stage</abbr>
<abbr>outputs</abbr> as plot or define plot configuration inside `dvc.yaml`
under `plots` key.

### Standalone plots

> Plots defined in `dvc.yaml` under the `plots` key are especially useful when
> users want to compare data from different data sources residing on the same
> version of the project. For example, comparing training versus test results on

Reviewer 3: ```suggestion
Plots can be defined in `dvc.yaml` under the `plots` key. Unlike
[stage plots](#stage-plots), these are especially useful when
users want to compare data from different data sources residing on the same
version of the project. For example, comparing training versus test results on
```

It might be better to split this into two sentences. Explain what standalone plots are, then explain their unique benefit in the next sentence (and reference stage plots since it's not otherwise clear what you are comparing them to).

    y:
      confusion_matrix_data.csv: predicted_class
    x: actual_class
    template: confusion
```

In this case we provided `confusion_matrix` as a plot id. It will be displayed
in the plot as a title, unless we override it with `title` field. In this case
we provided data source in `y` axis definition. Data will be sourced from
`confusion_matrix_data.csv`. As `y` axis we will use `predicted_class` field. On
`x` axis we will have `actual_class` field. Note that DVC will assume that
`actual_class` is inside `confusion_matrix_data.csv`.

We can provide multiple columns/fields from the same file:

```yaml
#dvc.yaml
---
plots:
  multiple_series:
    y:
      logs.csv: [accuracy, loss]
    x: epoch
```

In this case, we will take `accuracy` and `loss` fields and display them agains
`epoch` column, all coming from `logs.csv` file.

We can source the data from multiple files too:

```yaml
#dvc.yaml
---
plots:
  multiple_files:
    y:
      train_logs.csv: accuracy
      test_logs.csv: accuracy
    x: epoch
```

In this case we will plot `accuracy` field from both `train_logs.csv` and
`test_logs.csv` against the `epoch`. Note that both files have to have `epoch`
field.

> #### Avialable configuration fields

Reviewer 3: ```suggestion
#### Available configuration fields
```

- `x` - field name from which the X axis data comes from. An auto-generated
  _step_ field is used by default. It has to be a string.

- `y` - field name from which the Y axis data comes from. It can be a string,
  list or dictionary. If its a string or list, it is assumed that plot id will
  be the path to the data source. String, or list elements will be the names of
  data columns or fields withing the source file. If this field is a dictionary,
  it is assumed that its keys are paths to data sources. The values have to be

>   either strings or lists, and are treted as column(s)/field(s) within

Reviewer 3: ```suggestion
  either strings or lists, and are treated as column(s)/field(s) within
```

  respective files.
- `x_label` - X axis label. The X field name is the default.
- `y_label` - Y axis label. If all provided Y entries have the same field name,
  this name will be the default, `y` string otherwise.
- `title` - Plot title. Plot id by default.

Refer to the
[examples](/doc/command-reference/plots#example-simple-plot-definition) for more
syntax insight.

> ### Stage plots

Reviewer 3: WDYT about moving this above Syntax and then mentioning in Syntax when some options are specific to standalone plots definitions?

Author: Syntax is still part of plot configuration, so I would go with h3, to make it under `Configuring a plot`

When using `dvc run` or `dvc stage add`, instead of using
`--outs/--outs-no-cache` particular outputs can be marked with
`--plots/--plots-no-cache`. This will tell DVC that they are intended for
visualizations.

Upon running `dvc plots show/diff` DVC will collect plot outputs alongside the
definitions and display them conforming to their configuration. Note, that if
there are plot outputs in the project and they are also used in some plots
definitions, DVC will create separate rendering for the plot outputs and all

> definitions using them.

Reviewer 3: Let's consistently use one set of terms for the two types of plots in dvc.yaml. The headings use "standalone plots" and "stage plots," but most of the text uses "plot outputs" and "definitions." Let's pick one and apply it everywhere.

Author: Refactored.
Follow-up changes for help:
https://github.com/iterative/dvc/pull/7972

}
```

## Configuring a plot

In order to create visualizations, users need to provide the data and
(optionally) configuration that will help customize the plot. DVC provides two
ways to configure visualizations. Users can mark specific <abbr>stage</abbr>
<abbr>outputs</abbr> as plot or define plot configuration inside `dvc.yaml`
under `plots` key.

### Standalone plots

Plots can be defined in `dvc.yaml` under the `plots` key. Unlike
[stage plots](#stage-plots), these are especially useful when users want to
compare data from different data sources residing on the same version of the
project. For example, comparing training versus test results on current branch.

### Stage plots

> When using `dvc run` or `dvc stage add`, instead of using

Reviewer 4: ```suggestion
When using `dvc stage add`, instead of using
```

`--outs/--outs-no-cache` particular outputs can be marked with
`--plots/--plots-no-cache`. This will tell DVC that they are intended for
visualizations.

Upon running `dvc plots show/diff` DVC will collect stage plots alongside the
standalone plots and display them conforming to their configuration. Note, that
if there are stage plots in the project and they are also used in some plots
standalone plots, DVC will create separate rendering for the stage plots and all

> standalone plots using them.

Reviewer 4: I find this phrase a little confusing. Do we want to explain something like that the actual data files can be reused in the standalone plots and the stage plots?

Reviewer: Yeah the terminology is confusing. Issues I see:

**Configuring a plot**

Why is is separate from `plots modify` and **Plot templates**? Those are 2 actual ways to configure plots (already slightly confusing that there's 2 plot configs, let's avoid a 3rd). I think we're talking about "defining plots" here...

**Stand-alone plots**

They're linked to a stage via an `output` name though, right? So why stand-alone? I think this is about reusable, pipeline-level definitions.

**Stage plots**

Sounds like we're adding another dimension to the concept of "plot types" (dimension 1 being data-series vs. images) which may be confusing. And again, even pipeline-level plot definitions are linked to stages via an `output` name, so the term isn't accurate.

> Will try to post suggestions once I get a better understanding of the big picture...

Reviewer: OK here's what I'm thinking: This ref is now way too long. And even when clearly [we need to make a proper guide](https://github.com/iterative/dvc.org/issues/2572), a lot of the detail here can be compacated and reorganized. Specific plan of action I recommend:

**Configuring a plot** / **Stand-alone plots**/ **Stage plots**

Make this a paragraph or 2 in the existing **Description** of this page (ideally no H3 sections, or make a single **Defining plots** section if absolutely necessary). It probably replaces **Supported file formats** since that info is included in the plot definition and syntax blocks.

Terminology-wise, I'd go with descriptive phrasing like "plots defined in stages", "found in the `plots` field of `dvc.yaml`", etc. instead of coming up with special terms. From the user-perspective all plots are the same once rendered.

**Syntax (standalone plots only)**

Move all of this into [**Standalone plots configurations**](https://dvc-org-7086-flexible-p-d0hfxv.herokuapp.com/doc/user-guide/project-structure/pipelines-files#standalone-plots-configurations) (in the dvc.yaml schema guide) and link from this page as needed.

> I'd rename that section to **Reusable plot definitions** (will leave a specific suggestion ⏳)

Try to summarize as much as possible. Use tables if needed. I can jump in to help with that once the content is in the right place. Thanks

}
```

## Definining a plot

In order to create visualizations, users need to provide the data and
(optionally) configuration that will help customize the plot. DVC provides two
ways to configure visualizations. Users can mark specific <abbr>stage</abbr>
<abbr>outputs</abbr> as plot or define plot configuration inside `dvc.yaml`
under `plots` key.

> ## Plots definitions

Reviewer 3: Suggestion: Can the headers and descriptions better explain the two different ways to define plots? Maybe introduce some terminology like "Standalone Plots" and "Stage Plots"?

Reviewer 3: Also, should both this and the next header be subheadings of "Defining a plot"?

> Plots defined in `dvc.yaml` are especially useful when users want to compare

Reviewer 3: ```suggestion
Plots defined in `dvc.yaml` under the `plots` key are especially useful when users want to compare
```

> data from differend data sources residing on the same version of the project.

Reviewer 3: ```suggestion
data from different data sources residing on the same version of the project.
```

For example, comparing training versus test results on current branch.

### Syntax

> In order to define the plot users need to provide data and configuration for the

Author: No

plot. The plots should be defined in `dvc.yaml` file under `plots` key. Refer to
the [examples](/doc/command-reference/plots#example-simple-plot-definition) for
more syntax insight.

```yaml
# dvc.yaml
stages: ...

plots: ...
```

> ## Plot outputs

Reviewer 2: By reading this I don't understand how plot outputs are not related to the global "plots" section. Could we clarify this?

Reviewer 3: See also https://github.com/iterative/example-repos-dev/pull/117#issuecomment-1175197428

When using `dvc run` or `dvc stage add`, instead of using
`--outs/--outs-no-cache` particular outputs can be marked with
`--plots/--plots-no-cache`. This will tell DVC that they are intended for
visualizations. This special type of outputs might come in hand if users want to
visually compare experiments results with other experiments versions. For

> example, comparing new experiment with the baseline version of the project.

Reviewer 3: Why is this info under the section specific to plots outputs?

}
```

## Defining plots

In order to create visualizations, users need to provide the data and
(optionally) configuration that will help customize the plot. DVC provides two
ways to configure visualizations. Users can mark specific stage
<abbr>outputs</abbr> as plots or define top-level `plots` in `dvc.yaml`.

### Stage plots

When using `dvc stage add`, instead of using `--outs/--outs-no-cache` particular
outputs can be marked with `--plots/--plots-no-cache`. This will tell DVC that
they are intended for visualizations.

Upon running `dvc plots show/diff` DVC will collect stage plots alongside the
[top-level plots](#top-level-plots) and display them conforming to their
configuration. Note, that if there are stage plots in the project and they are
also used in some top-level definitions, DVC will create separate rendering for
the stage plots and all definitions using them.

> This special type of outputs might come in hand if users want to visually

Reviewer 3: ```suggestion
This special type of outputs might come in handy if users want to visually
```

compare experiments results with other experiments versions and not bother with
writing top-level plot definitions in `dvc.yaml`.

### Top-level plots

Plots can also be defined in a top-level `plots` key in `dvc.yaml`. Unlike
[stage plots](#stage-plots), these definitions let you overlay plots from
different data sources, for example training vs. test results (on the current
project version). Conversely, you can create multiple plots from a single source

> file. You can also any plot file in the project, regardless of whether it's a
> stage outputs. This creates a separation between visualization and outputs.

Reviewer 3: ```suggestion
file. You can also use any plot file in the project, regardless of whether it's a
stage outputs. This creates a separation between visualization and outputs.
```

In order to define the plot users need to provide data and an optional
configuration for the plot. The plots should be defined in `dvc.yaml` file under
`plots` key.

```yaml
# dvc.yaml
stages: ...

plots: ...
```

> Every plots has to have its own ID. Configuration, if provided, should be a
> dictionary.
> 
> In simplest use case, user can provide file path as the plot ID and don't
> provide configuration at all:

Reviewer 3: ```suggestion
Every plot has to have its own ID. Configuration, if provided, should be a
dictionary.

In the simplest use case, a user can provide the file path as the plot ID and not
provide configuration at all:
```

}
```

## Configuring a plot

In order to create visualizations, users need to provide the data and
(optionally) configuration that will help customize the plot. DVC provides two
ways to configure visualizations. Users can mark specific <abbr>stage</abbr>
<abbr>outputs</abbr> as plot or define plot configuration inside `dvc.yaml`
under `plots` key.

### Top-level plot definitions

Plots can be defined in `dvc.yaml` under the `plots` key. Unlike
[stage plots](#stage-plots), these are especially useful when users want to
compare data from different data sources residing on the same version of the

> project. For example, comparing training versus test results on current branch.

Reviewer: > They enable covering **multiple data sources in one plot** or **using the same data source in multiple plots**...
> The main point is that they can **use data sourced from different files**, maybe we should include that?

Also found more specific context in https://github.com/iterative/dvc/issues/7086#issuecomment-985640176:
> 1. iterative/dvc#6316 (single-source segmentation)
> 2. Multiple series per axis (single-source comparison)
> 3. Multiple files within one plot (multi-source composite)

... Any more?

We should expand a bit on the benefits/features of top-level definitions (which may soonish become the only documented place to define plots per https://github.com/iterative/dvc.org/issues/3763). "to compare data from different data sources residing on the same version of the project" doesn't seem to cover all of it.

Author: I think that this one sentence pretty much covers all of it, but I agree that we might not be doing the best job explaining its value here. Let me think how to modify that.

Reviewer 3: @jorgeorpinel As you mention above, we shouldn't be comparing to "stage plots" if the "top-level" ones are the default and soon the only supported method. Do we need to explain their unique benefits in that case? Maybe we instead need to explain "stage plots" as a subset that can be defined inside a pipeline stage but are limited to configuring a single output as a plot?

More importantly, should we continue to discuss in this PR or is the rest of it in good enough shape that we could merge and iterate on top?

Reviewer: > > compare data from different data sources
>
> this one sentence pretty much covers all of it

@pared what about the same data source file for 2+ plots?

> Do we need to explain their unique benefits?

@dberenbaum right, I didn't mean the benefits over stage-level plots but rather just the benefits/features of plot definitions (which are top-level).

> explain "stage plots" as a subset

Best not to worry about updating those existing explanations until #3767 is addressed. This PR is already complex.

> should we continue to discuss in this PR or is the rest of it in good enough shape

This question shouldn't a blocker but we should probably make a summary of pending questions and try to get this approved... ⌛

Author: > @pared what about the same data source file for 2+ plots?

You got me here, yes same data source file can also be used

  train:
    cmd: echo "Hello world"

plots:
  logs.csv:
    x: epoch
    y: accuracy
    title: Displaying accuracy
    x_label: This is epoch
    y_label: This is accuracy
```

```dvc
$ dvc plots show
file:///Users/usr/src/dvc_plots/index.html
```

![](/img/plots_show_spec_simple_custom.svg)

## Example: multiple data-series plot definition:

Data in `training_data.csv`:

```csv
epoch,train_loss,test_loss
1,0.33,0.4
2,0.3,0.28
3,0.2,0.25
4,0.1,0.23
```

```yaml
# dvc.yaml
stages:
  train:
    cmd: echo "Hello world"

plots:
  test_vs_train_loss:
    x: epoch
    y:
      training_data.csv: [test_loss, train_loss]
    title: Compare loss training versus test
```

![](/img/plots_show_spec_multiple_columns.svg)

## Example: sourcing data from different files

Lets prepare comparison for confusion matrix data between test set and training
set:

```csv
# train_classes.csv
actual_class,predicted_class
dog,dog
dog,dog
dog,dog
dog,bird
cat,cat
cat,cat
cat,cat
cat,dog
bird,bird
bird,bird
bird,bird
bird,dog
```

```csv
# test_classes.csv
actual_class,predicted_class
dog,dog
dog,dog
dog,cat
bird,bird
bird,bird
bird,cat
cat,cat
cat,cat
cat,bird
```

```yaml
# dvc.yaml
stages:
  train:

>     cmd: echo "Hello world"

Reviewer 2: make it more realistic? or remove completely (we don't anything like this in all the other cases)

Reviewer 2: it makes reproducible though ...

make it "Training a model ... " at least?

Reviewer: I think in ref examples realistic > reproducible. We don't expect readers to run the examples as in a tutorial right? They're mostly to see them, maybe copy the code but change it to fit their project.

Author: Modified.

plots:
  test_vs_train_confusion:
    x: actual_class
    y:
      train_classes.csv: predicted_class
      test_classes.csv: predicted_class
    title: Compare test vs train confusion matrix

>     template: confusion

Reviewer 2: how does diff look like in this case (unmergable plots)? can you show an example please? just trying to understand the consequences for VS Code / Studio

Reviewer 2: Got it. I would say it's confusing to see `rev` as a title a bit. Not related to this PR obviously.

I think it's fine to have a simple case for these plots. I don't it's required or something.

Author: Yep, I will have to fix that. Regretfully, because in the beginning, I did not convert `rev` into an anchor, this field will have to stay with us for some time. I should probably be able to remove it from the subtitle though.

Reviewer 3: Do we think it's useful to include a diff example here (doesn't need to block the PR)?

- `linear` - basic linear plot including cursor interactivity (default)
- `simple` - simplest linear template (not interactive); Good base to create

>   [custom template].
> - `scatter` - scatter plot
> - `smooth` - linear plot with LOESS smoothing, see
>   [example](/doc/command-reference/plots#example-smooth-plot)

Reviewer: Should be plural (1/2)
```suggestion
  [custom templates].
- `scatter` - scatter plot
- `smooth` - linear plot with LOESS smoothing, see
  [example](/doc/command-reference/plots#example-smooth-plot)
```

Reviewer: Shoot I can't easily fix the actual link in line 271... (should be `[custom templates]` there)... ⏳

Reviewer: OK it's time for me to clone this fork 😬 ...

Reviewer: Fixed.

![](/img/plots_show_confusion.svg)

> A confusion matrix [template](/doc/command-reference/plots#plot-templates) is
> predefined in DVC.

We can use `confusion_normalized` template to normalize the results:

```dvc
$ dvc plots show classes.csv --template confusion_normalized -x actual -y predicted
file:///Users/usr/src/dvc_plots/index.html
```

![](/img/plots_show_confusion_normalized.svg)

> ## Example: simple plot definition

Reviewer 2: My feeling that we should create a large table with three-four columns - data, template, spec (dvc.yaml), image - and put a lot of examples there - from some basic regular things, to more advanced. So that people can quickly start, search for it, copy paste, etc

Reviewer 3: Some more organization would definitely be useful, and more examples would be great.

For now, could we at least try to break it up so it's not one huge list of examples? Maybe have different subsections. It seems like there's a bunch of path-based examples, followed by dvc.yaml examples, and then finally a custom html template example.

Let's get back to the `logs.csv` data:

```
# logs.csv
epoch,loss,accuracy
1,0.19,0.81
2,0.11,0.89
3,0.07,0.93
4,0.04,0.96
```

Minimal plot definition we can put in `dvc.yaml` is simply data source path
relative to `dvc.yaml` file:

```yaml
# dvc.yaml
stages:
  train:
    cmd: echo "Training the model..."

plots:
  logs.csv:
```

```dvc
$ dvc plots show
file:///Users/usr/src/dvc_plots/index.html
```

![](/img/plots_show_spec_default.svg)

We can customize it:

```yaml
# dvc.yaml
stages:
  train:
    cmd: echo "Training the model..."

plots:
  logs.csv:
    x: epoch
    y: accuracy
    title: Displaying accuracy
    x_label: This is epoch
    y_label: This is accuracy
```

```dvc
$ dvc plots show
file:///Users/usr/src/dvc_plots/index.html
```

![](/img/plots_show_spec_simple_custom.svg)

> ## Example: multiple data-series plot definition:

Reviewer 3: ```suggestion
## Example: multiple data-series plot definition
```

Data in `training_data.csv`:

```csv
epoch,train_loss,test_loss
1,0.33,0.4
2,0.3,0.28
3,0.2,0.25
4,0.1,0.23
```

```yaml
# dvc.yaml
stages:
  train:
    cmd: echo "Training the model..."

plots:
  test_vs_train_loss:
    x: epoch
    y:
      training_data.csv: [test_loss, train_loss]
    title: Compare loss training versus test
```

> 

Reviewer 3: `dvc plots show`?

![](/img/plots_show_spec_multiple_columns.svg)

## Example: sourcing data from different files

Lets prepare comparison for confusion matrix data between test set and training
set:

```csv
# train_classes.csv
actual_class,predicted_class
dog,dog
dog,dog
dog,dog
dog,bird
cat,cat
cat,cat
cat,cat
cat,dog
bird,bird
bird,bird
bird,bird
bird,dog
```

```csv
# test_classes.csv
actual_class,predicted_class
dog,dog
dog,dog
dog,cat
bird,bird
bird,bird
bird,cat
cat,cat
cat,cat
cat,bird
```

```yaml
# dvc.yaml
stages:
  train:
    cmd: echo "Training the model..."

plots:
  test_vs_train_confusion:
    x: actual_class
    y:
      train_classes.csv: predicted_class
      test_classes.csv: predicted_class
    title: Compare test vs train confusion matrix
    template: confusion
    x_label: Actual class
    y_label: Predicted class
```

> 

Reviewer 3: `dvc plots show`?

- `-v`, `--verbose` - displays detailed tracing information.

## Examples

### Raw data files

#### Tabular data

> We'll use tabular data file `logs.csv` for this example:

Author: Moved the examples to respective subcommands.

> We'll use tabular data file `logs.csv` for this example:

Reviewer: Thanks @pared

![](/img/plots_show_confusion.svg)

> A confusion matrix [template](/doc/command-reference/plots#plot-templates) is
> predefined in DVC.

We can use `confusion_normalized` template to normalize the results:

```dvc
$ dvc plots show classes.csv --template confusion_normalized -x actual -y predicted
file:///Users/usr/src/dvc_plots/index.html
```

![](/img/plots_show_confusion_normalized.svg)

### Top-level plots

#### Simple plot definition

> Let's get back to the `logs.csv` data:

Reviewer: I'm not sure we want many examples if at all in the `plots` ref index page actually. Shouldn't they all be in the sub-refs? Unless a relevant example combines more than one sub-command (in this case I only see `plots show`).

Or if it's about different ways to define plots, that could be summarized into the `dvc.yaml` schema guide.

Author: I guess we should move the examples to `show` `diff`. The syntax is already explained in `Syntax`
